phosphobot/models.py

```diff
@@ -1060,6 +1076,157 @@ def save(self, meta_folder_path: str) -> None:
 
         self.to_json(meta_folder_path)
 
+    def update_before_episode_removal(self, parquet_path: str) -> None:
+        """
+        Update the stats before removing an episode from the dataset.
+        We do not compute the new min and max.
+        We prefer to do it after the episode removal to directly access new indexes and episodes indexes.
+        """
+        # Check the parquet file exists
+        if not os.path.exists(parquet_path):
+            raise ValueError(f"Parquet file {parquet_path} does not exist")
+
+        data_folder_path = "/".join(parquet_path.split("/")[:-1])
+
+        # Load the parquet file
+        deleted_episode_df = pd.read_parquet(parquet_path)
+        nb_steps_deleted_episode = len(deleted_episode_df)
+
+        # For each column in the parquet file, compute sum along first axis, min/max along last axis
+        logger.info("Updating stats before removing episode")
+        logger.info(f"Episode df action: {deleted_episode_df['action']}")
+
+        # Compute the sum and square sum for each column
+        column_sums = {
+            col: {
+                "sum": np.sum(np.array(deleted_episode_df[col].tolist()), axis=0),
+                "square_sum": np.sum(
+                    np.array(deleted_episode_df[col].tolist()) ** 2, axis=0
+                ),
+            }
+            for col in deleted_episode_df.columns
+        }
+        logger.info(f"Column sums: {column_sums}")
+        # Update stats for each field in the StatsModel
+        for field_name, field in StatsModel.model_fields.items():
+            # TODO task_index is not updated since we do not support multiple tasks
+            # observation_images has a special treatment
+            if field_name in ["task_index", "observation_images"]:
+                continue
+            logger.info(f"Updating field {field_name}")
+            # Get the field value from the instance
+            field_value = getattr(self, field_name)
+            # Convert observation_state to observation.state
+            field_name = (
+                "observation.state" if field_name == "observation_state" else field_name
+            )
+            # Update statistics
+            if field_name in column_sums:
+                # Maximum of debug info
+                logger.info(f"Column sums for {field_name}: {column_sums[field_name]}")
+                logger.info(f"Field value: {field_value}")
+                logger.info(f"Field value mean: {field_value.mean}")
+
+                # Subtract sums
+                field_value.sum -= column_sums[field_name]["sum"]
+                field_value.square_sum -= column_sums[field_name]["square_sum"]
```

Comment:
please do : field_value.sum = field_value.sum - column_sums[field_name]["sum"] to avoid pointer errors